flocker/control/_persistence.py

```diff
@@ -21,16 +21,94 @@
 from twisted.internet.defer import succeed
 from twisted.internet.task import LoopingCall
 
-from ._model import SERIALIZABLE_CLASSES, Deployment
+from ._model import SERIALIZABLE_CLASSES, Deployment, Configuration
 
 
 # Serialization marker storing the class name:
 _CLASS_MARKER = u"$__class__$"
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CONFIG_VERSION = 2
 
-class _ConfigurationEncoder(JSONEncoder):
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to take place
+    or complete successfully.
+    """
+
+
+def migrate_configuration(source_version, target_version,
+                          config, migration_class):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``ConfigurationMigration`` class methods for
+    the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
```

Comment:
Alternatively, we could suppose that we're going to stick with JSON (at least for a long while) and do the json encoding/decoding outside of the upgrade methods.  The upgrade methods then get to deal with decoded objects and don't have to all replicate the same json encode/decode steps.